wgsl/index.bs

```diff
@@ -989,86 +988,365 @@ Host-shareable type |S| satisfies <dfn noexport>uniform buffer layout</dfn> when
+the resulting pointer always remains in bounds of the original variable,
+i.e. the resulting pointer can only reference storage within the storage allocated
+to the originating variable.
+There are two cases:
+* When the expression uses a fixed name to form a pointer to a sub-part of a composite value,
+    the [SHORTNAME] program is invalid if the name is out of range.
+    * <p algorithm="letter access out of bounds example">
+        For example, if |v| is a pointer to a vector with fewer than four elements,
+        then the program is invalid if it has the expression |v|`.w`.
+        </p>
+* When the expression uses an array-index phrase to form a pointer to
+    a vector component, a matrix column, or an array element,
+    then [SHORTNAME] program is statically valid.
+    However at runtime, if the index value is out of range for the type,
+    or out of range for the runtime size of a runtime-sized array,
+    then an in-bounds index value will be used instead.
+    * <p algorithm="indexed access out of bounds example">
+        For example, if |array_var| is a pointer to an array of 10 elements, then
+        |array_var|`[`|i|`]` is a valid expression.
+        However, if at runtime the value of |i| is negative or larger than 9,
+        then it is out of bounds and instead the resulting
+        element pointer will be formed using an index value in the range [0, 9].
+        </p>
 
-Note: Assignment of swizzled values is not permitted (SubaccessSwizzle).<br>
-           e.g. `vec4<i32> v; v.xz = vec2<i32>(0, 1);` is not allowed.
+See details at [[#composite-value-decomposition-expr]].
+
+### Accessing memory via a pointer ### {#accessing-memory-via-a-pointer}
+
+Pointers are used to write into storage, and to read from storage:
+
+: <dfn noexport>Store Rule</dfn>:
+:: An assignment statement stores the value on its right hand side into the storage referenced by its left hand side.
+    See [[#assignment]].
+: <dfn noexport>Load Rule</dfn>:
+:: When a pointer expression is used as a subexpression, and the type rules of [SHORTNAME]
+    require the subexpression to have a non-pointer type, then:
+    * The non-pointer type must be the same as the [=pointee type=] of the pointer.
+    * The resulting value of the expression is the value currently stored in the storage
+        referenced by the pointer.
+        In other words, the evaluation causes the memory locations to be read (loaded),
+        and yields the value that was read.
+
+Note: The [=load rule=] is applied only if type checking would otherwise fail,
+i.e. the application
+of *other* type rules would necessarily lead to an inability to deduce a valid type for
+the all expressions in the statement.
+
+**Important:**
+Pointer expressions often look the same as non-pointer expressions.
+However, the [=load rule=] is only applied as a last resort: if it is possible
+to interpret the expression so that it yields a pointer, then that is how the expression
```

Comment:
I guess this load rule is what confuses me most. I believe this complexity is unnecessary, and pointer construction could be more explicit, without ruining the type inference.

Reply:
You pay either like this, or by not knowing if you have an l-value or an r-value.

Reply:
I wonder if l-values are going to be a problem for us, really. Unlike C, we don't need the "=" to produce a result, we don't need to support "a = b = c", etc. So for WGSL, supporting syntax like "*pointer = value;" could be trivial:
  - in the statement context, if it starts with "*", we are looking at the assignment (i.e. the l-value would follow).

Are there hidden pitfalls?

Reply:
Small update: I we are *already* tracking L-values versus R-values in WGSL frontend. I'd assume Tint does so, as well. I.e. when you parse "something.other = value", the parser needs to understand that "something.other" is an L-value. So de-referencing pointers with "*" doesn't introduce any more complexity in here than we already have.